File: darkmulti/commands.py
"""Command implementations for dark-multi."""

import shutil
import time

from .config import (
    DARK_ROOT, PROXY_PORT,
    get_system_resources, suggest_max_instances,
    Colors, log, error, success, warn, run
)
from .branch import Branch, find_next_instance_id, find_source_repo, get_managed_branches
from .tmux import Tmux
from .proxy import is_proxy_running, start_proxy_server, stop_proxy_server, ensure_proxy_running
from .dns import setup_dns
from .devcontainer import get_override_config_path, generate_override_config


def cmd_ls(args) -> int:
    """List all branches."""
    cpu_cores, ram_gb = get_system_resources()
    suggested = suggest_max_instances()

    print(f"Branches in {DARK_ROOT}:")
    print(f"  System: {cpu_cores} cores, {ram_gb}GB RAM -> suggested max: {suggested} concurrent\n")

    branches = get_managed_branches()
    running_count = sum(1 for b in branches if b.is_running)

    if branches:
        for b in branches:
            print(f"  {b.status_line()}")
        print(f"\n  Running: {running_count}/{suggested} suggested max")
    else:
        print("  (no branches)")
        print("\n  Create one: multi new <branch>")

    print()
    if Tmux.session_exists():
        print(f"tmux session 'dark' exists. Attach: multi")
    else:
        print("No tmux session yet.")

    return 0


def cmd_new(args) -> int:
    """Create a new branch clone."""
    name = args.branch
    base = args.base

    branch = Branch(name)

    if branch.exists:
        error(f"Branch '{name}' already exists at {branch.path}")
        return 1

    source = find_source_repo()
    if not source:
        error("No source repo found. Set DARK_SOURCE or create 'main' first.")
        return 1

    instance_id = find_next_instance_id()

    log(f"Creating branch '{name}' from {source}")
    log(f"  Instance ID: {instance_id}, ports: {10011 + instance_id * 100}+")

    DARK_ROOT.mkdir(parents=True, exist_ok=True)

    # Clone
    result = run(["git", "clone", str(source), str(branch.path)], capture_output=False)
    if result.returncode != 0:
        error("Clone failed")
        return 1

    # Setup branch
    log(f"Checking out branch '{name}' from '{base}'...")
    run(["git", "fetch", "origin"], cwd=branch.path)
    # Try tracking remote branch, or create new
    result = run(["git", "checkout", "-b", name, f"origin/{base}"], cwd=branch.path)
    if result.returncode != 0:
        run(["git", "checkout", "-b", name, base], cwd=branch.path)

    # Write metadata
    branch.write_metadata(instance_id)

    # Generate override config (keeps repo's devcontainer.json untouched)
    override_path = generate_override_config(branch)
    log(f"Generated override config at {override_path}")

    # Start container
    log("Starting devcontainer...")
    if not shutil.which("devcontainer"):
        error("devcontainer CLI not found. Install: npm install -g @devcontainers/cli")
        return 1

    result = run([
        "devcontainer", "up",
        "--workspace-folder", str(branch.path),
        "--override-config", str(override_path)
    ], capture_output=False)
    if result.returncode != 0:
        error("Failed to start devcontainer")
        return 1

    time.sleep(2)

    container_id = branch.container_id
    if not container_id:
        error("Container started but couldn't find it")
        return 1

    # Create tmux window
    log("Setting up tmux window...")
    Tmux.create_window(name, container_id, branch.path)

    # Open VS Code unless --no-code
    if not args.no_code:
        log("Opening VS Code...")
        open_vscode(branch)

    # Ensure proxy is running for URL access
    ensure_proxy_running()

    success(f"Branch '{name}' ready!")
    print(f"\nAttach tmux: multi")
    print(f"URLs: multi urls")
    return 0


def cmd_start(args) -> int:
    """Start a stopped branch."""
    name = args.branch
    branch = Branch(name)

    if not branch.exists:
        error(f"Branch '{name}' not found. Create it: multi new {name}")
        return 1

    if not branch.is_managed:
        error(f"'{name}' is not a managed branch")
        return 1

    if branch.is_running:
        warn(f"Branch '{name}' already running")
        if not Tmux.window_exists(name):
            log("Adding tmux window...")
            Tmux.create_window(name, branch.container_id, branch.path)
        return 0

    log(f"Starting branch '{name}'...")

    if not shutil.which("devcontainer"):
        error("devcontainer CLI not found")
        return 1

    # Ensure override config exists
    override_path = get_override_config_path(branch)
    if not override_path.exists():
        override_path = generate_override_config(branch)
        log(f"Generated override config at {override_path}")

    result = run([
        "devcontainer", "up",
        "--workspace-folder", str(branch.path),
        "--override-config", str(override_path)
    ], capture_output=False)
    if result.returncode != 0:
        error("Failed to start devcontainer")
        return 1

    time.sleep(2)

    container_id = branch.container_id
    if not container_id:
        error("Container started but couldn't find it")
        return 1

    log("Setting up tmux window...")
    Tmux.create_window(name, container_id, branch.path)

    # Open VS Code unless --no-code
    if not args.no_code:
        log("Opening VS Code...")
        open_vscode(branch)

    # Ensure proxy is running for URL access
    ensure_proxy_running()

    success(f"Branch '{name}' running")
    return 0


def cmd_stop(args) -> int:
    """Stop a branch (keeps files)."""
    name = args.branch
    branch = Branch(name)

    if not branch.exists:
        error(f"Branch '{name}' not found")
        return 1

    log(f"Stopping branch '{name}'...")

    # Kill tmux window
    Tmux.kill_window(name)

    # Stop and remove container
    container_id = branch.container_id
    if container_id:
        log("Stopping container...")
        run(["docker", "stop", container_id])
        run(["docker", "rm", container_id])

    success(f"Branch '{name}' stopped. Files at {branch.path}")
    return 0


def cmd_rm(args) -> int:
    """Remove a branch entirely with full cleanup."""
    from .config import CONFIG_DIR

    name = args.branch
    branch = Branch(name)

    if not branch.exists:
        error(f"Branch '{name}' not found")
        return 1

    if not branch.is_managed:
        error(f"'{name}' is not a managed branch")
        return 1

    # Confirmation
    if not args.force:
        if branch.has_changes:
            warn(f"Branch '{name}' has uncommitted changes!")

        print(f"This will remove branch '{name}':")
        print(f"  - Stop and remove container")
        print(f"  - Remove tmux window")
        print(f"  - Delete {branch.path}")
        response = input("Proceed? [y/N] ").strip().lower()
        if response != "y":
            print("Aborted")
            return 1

    log(f"Removing branch '{name}'...")

    # 1. Kill tmux window
    Tmux.kill_window(name)

    # 2. Stop and remove container
    container_id = branch.container_id
    if container_id:
        log("Stopping container...")
        run(["docker", "stop", container_id])
        run(["docker", "rm", container_id])

    # 3. Clean up any dangling containers with this label (stopped ones)
    result = run(["docker", "ps", "-aq", "--filter", f"label=dark-dev-container={name}"])
    for cid in result.stdout.strip().split("\n"):
        if cid:
            run(["docker", "rm", "-f", cid])

    # 4. Remove override config
    override_dir = CONFIG_DIR / "overrides" / name
    if override_dir.exists():
        shutil.rmtree(override_dir)

    # 5. Remove directory
    log("Removing files...")
    shutil.rmtree(branch.path)

    success(f"Branch '{name}' removed")
    return 0


def open_vscode(branch: Branch) -> bool:
    """Open VS Code for a branch. Returns True on success."""
    if not branch.is_running:
        return False

    # Use devcontainer CLI to open VS Code
    if shutil.which("devcontainer"):
        result = run(["devcontainer", "open", str(branch.path)], capture_output=False)
        if result.returncode == 0:
            return True

    # Fallback: use code --remote
    if shutil.which("code"):
        import binascii
        container_id = branch.container_id
        hex_id = binascii.hexlify(container_id.encode()).decode()
        result = run(["code", "--remote", f"attached-container+{hex_id}", "/home/dark/app"], capture_output=False)
        return result.returncode == 0

    warn("Neither devcontainer CLI nor VS Code found")
    return False


def cmd_code(args) -> int:
    """Open VS Code attached to branch container."""
    name = args.branch
    branch = Branch(name)

    if not branch.exists:
        error(f"Branch '{name}' not found")
        return 1

    if not branch.is_running:
        error(f"Branch '{name}' not running. Start it first: multi start {name}")
        return 1

    log(f"Opening VS Code for '{name}'...")
    if open_vscode(branch):
        return 0
    else:
        error("Failed to open VS Code")
        return 1


def cmd_attach(args) -> int:
    """Attach to tmux session."""
    branches = get_managed_branches()

    if not branches:
        cpu_cores, ram_gb = get_system_resources()
        suggested = suggest_max_instances()

        print()
        print("No branches found. Let's set up!")
        print()
        print(f"System: {cpu_cores} cores, {ram_gb}GB RAM")
        print(f"Suggested max concurrent: {suggested}")
        print()
        print("To get started:")
        print("  1. Set DARK_SOURCE if your dark repo isn't at ~/code/dark:")
        print("     export DARK_SOURCE=~/code/dark-source")
        print()
        print("  2. Create your first branch:")
        print("     multi new main")
        print()
        return 1

    # Auto-create tmux windows for running branches
    running_branches = [b for b in branches if b.is_running]
    if running_branches:
        # Ensure meta window exists first
        Tmux.ensure_meta_window()
        for b in running_branches:
            if not Tmux.window_exists(b.name):
                log(f"Creating tmux window for {b.name}...")
                Tmux.create_window(b.name, b.container_id, b.path)
    else:
        print("No running branches. Start one first:")
        print("  multi start <branch>")
        print()
        cmd_ls(args)
        return 0

    Tmux.attach()
    return 0


def cmd_proxy(args) -> int:
    """Start/stop/status the URL proxy server."""
    action = args.action

    if action == "start":
        pid = is_proxy_running()
        if pid:
            warn(f"Proxy already running (PID {pid})")
            return 0

        log(f"Starting proxy on port {PROXY_PORT}...")
        pid = start_proxy_server(PROXY_PORT, background=True)
        success(f"Proxy started (PID {pid})")
        print()
        print(f"For browser access, run: multi setup-dns (one-time)")
        print(f"Then visit: http://dark-packages.<branch>.dlio.localhost:{PROXY_PORT}/ping")
        return 0

    elif action == "stop":
        if stop_proxy_server():
            success("Proxy stopped")
        else:
            warn("Proxy not running")
        return 0

    elif action == "status":
        pid = is_proxy_running()
        if pid:
            print(f"Proxy running (PID {pid}) on port {PROXY_PORT}")
        else:
            print("Proxy not running")
        return 0

    elif action == "fg":
        # Run in foreground (for debugging)
        pid = is_proxy_running()
        if pid:
            warn(f"Proxy already running in background (PID {pid})")
            return 1
        log(f"Starting proxy on port {PROXY_PORT} (foreground)...")
        start_proxy_server(PROXY_PORT, background=False)
        return 0

    return 1


def cmd_urls(args) -> int:
    """List available URLs for running branches."""
    branches = get_managed_branches()
    running = [b for b in branches if b.is_running]

    if not running:
        print("No running branches.")
        return 0

    proxy_pid = is_proxy_running()
    proxy_status = f"running (PID {proxy_pid})" if proxy_pid else "not running"

    print(f"Proxy: {proxy_status}")
    print(f"Port: {PROXY_PORT}")
    print()

    print("Running branches:")
    for b in running:
        print()
        print(f"  {Colors.BOLD}{b.name}{Colors.NC} (ID={b.instance_id})")
        print(f"    Direct:  curl -H 'Host: dark-packages.dlio.localhost' http://localhost:{b.bwd_port_base}/ping")
        if proxy_pid:
            print(f"    Proxy:   http://dark-packages.{b.name}.dlio.localhost:{PROXY_PORT}/ping")
        print(f"    BwdServer ports: {b.bwd_port_base}-{b.bwd_port_base + 1}")
        print(f"    Test ports: {b.port_base}-{b.port_base + 19}")

    if not proxy_pid:
        print()
        print("Start proxy for nice URLs: multi proxy start")

    print()
    print("For browser access, run: multi setup-dns (one-time)")

    return 0


def cmd_setup_dns(args) -> int:
    """Set up wildcard DNS for *.dlio.localhost -> 127.0.0.1"""
    return setup_dns()

File: internal/cli/commands.go
// Package cli provides CLI commands for dark-multi.
package cli

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"time"

	"github.com/spf13/cobra"

	"github.com/darklang/dark-multi/internal/branch"
	"github.com/darklang/dark-multi/internal/config"
	"github.com/darklang/dark-multi/internal/container"
	"github.com/darklang/dark-multi/internal/dns"
	"github.com/darklang/dark-multi/internal/proxy"
	"github.com/darklang/dark-multi/internal/tmux"
	"github.com/darklang/dark-multi/internal/tui"
)

// NewRootCmd creates the root cobra command.
func NewRootCmd() *cobra.Command {
	rootCmd := &cobra.Command{
		Use:   "multi",
		Short: "Manage multiple Dark devcontainer instances",
		Long: `Dark Multi - Manage multiple Dark devcontainer instances with tmux integration.

Examples:
  multi                      Attach to tmux session (or interactive TUI)
  multi ls                   List branches and status
  multi new fix-parser       Create new branch from main
  multi start fix-parser     Start stopped branch
  multi stop fix-parser      Stop branch (keeps files)
  multi rm fix-parser        Remove branch entirely
  multi code fix-parser      Open VS Code for branch
  multi urls                 List available URLs
  multi proxy start          Start URL proxy server
  multi setup-dns            Set up wildcard DNS`,
		Run: func(cmd *cobra.Command, args []string) {
			// Default action: attach to tmux or show TUI
			cmdAttach(cmd, args)
		},
	}

	rootCmd.AddCommand(lsCmd())
	rootCmd.AddCommand(newCmd())
	rootCmd.AddCommand(startCmd())
	rootCmd.AddCommand(stopCmd())
	rootCmd.AddCommand(rmCmd())
	rootCmd.AddCommand(codeCmd())
	rootCmd.AddCommand(urlsCmd())
	rootCmd.AddCommand(proxyCmd())
	rootCmd.AddCommand(setupDNSCmd())

	return rootCmd
}

func lsCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "ls",
		Short: "List branches",
		Run: func(cmd *cobra.Command, args []string) {
			cpuCores, ramGB := config.GetSystemResources()
			suggested := config.SuggestMaxInstances()

			fmt.Printf("Branches in %s:\n", config.DarkRoot)
			fmt.Printf("  System: %d cores, %dGB RAM -> suggested max: %d concurrent\n\n", cpuCores, ramGB, suggested)

			branches := branch.GetManagedBranches()
			runningCount := 0
			for _, b := range branches {
				if b.IsRunning() {
					runningCount++
				}
			}

			if len(branches) > 0 {
				for _, b := range branches {
					status := "\033[0;31mstopped\033[0m"
					if b.IsRunning() {
						status = "\033[0;32mrunning\033[0m"
					}
					changes := ""
					if b.HasChanges() {
						changes = " \033[1;33m[modified]\033[0m"
					}
					fmt.Printf("  \033[1m%-20s\033[0m %s  ports %d+/%d+%s\n",
						b.Name, status, b.PortBase(), b.BwdPortBase(), changes)
				}
				fmt.Printf("\n  Running: %d/%d suggested max\n", runningCount, suggested)
			} else {
				fmt.Println("  (no branches)")
				fmt.Println("\n  Create one: multi new <branch>")
			}

			fmt.Println()
			if tmux.SessionExists() {
				fmt.Printf("tmux session '%s' exists. Attach: multi\n", config.TmuxSession)
			} else {
				fmt.Println("No tmux session yet.")
			}
		},
	}
}

func newCmd() *cobra.Command {
	var base string
	var noCode bool

	cmd := &cobra.Command{
		Use:   "new <branch>",
		Short: "Create new branch",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			name := args[0]
			b := branch.New(name)

			if b.Exists() {
				fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m Branch '%s' already exists at %s\n", name, b.Path)
				os.Exit(1)
			}

			source := branch.FindSourceRepo()
			if source == "" {
				fmt.Fprintln(os.Stderr, "\033[0;31merror:\033[0m No source repo found. Set DARK_SOURCE or create 'main' first.")
				os.Exit(1)
			}

			instanceID := branch.FindNextInstanceID()

			fmt.Printf("\033[0;34m>\033[0m Creating branch '%s' from %s\n", name, source)
			fmt.Printf("\033[0;34m>\033[0m   Instance ID: %d, ports: %d+\n", instanceID, 10011+instanceID*100)

			os.MkdirAll(config.DarkRoot, 0755)

			// Clone
			cloneCmd := exec.Command("git", "clone", source, b.Path)
			cloneCmd.Stdout = os.Stdout
			cloneCmd.Stderr = os.Stderr
			if err := cloneCmd.Run(); err != nil {
				fmt.Fprintln(os.Stderr, "\033[0;31merror:\033[0m Clone failed")
				os.Exit(1)
			}

			// Setup branch
			fmt.Printf("\033[0;34m>\033[0m Checking out branch '%s' from '%s'...\n", name, base)
			exec.Command("git", "-C", b.Path, "fetch", "origin").Run()
			checkoutCmd := exec.Command("git", "-C", b.Path, "checkout", "-b", name, "origin/"+base)
			if err := checkoutCmd.Run(); err != nil {
				exec.Command("git", "-C", b.Path, "checkout", "-b", name, base).Run()
			}

			// Write metadata
			b.WriteMetadata(instanceID)

			// Generate override config
			overridePath, err := container.GenerateOverrideConfig(b)
			if err != nil {
				fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m Failed to generate config: %v\n", err)
				os.Exit(1)
			}
			fmt.Printf("\033[0;34m>\033[0m Generated override config at %s\n", overridePath)

			// Start container
			fmt.Println("\033[0;34m>\033[0m Starting devcontainer...")
			if _, err := exec.LookPath("devcontainer"); err != nil {
				fmt.Fprintln(os.Stderr, "\033[0;31merror:\033[0m devcontainer CLI not found. Install: npm install -g @devcontainers/cli")
				os.Exit(1)
			}

			devCmd := exec.Command("devcontainer", "up", "--workspace-folder", b.Path, "--override-config", overridePath)
			devCmd.Stdout = os.Stdout
			devCmd.Stderr = os.Stderr
			if err := devCmd.Run(); err != nil {
				fmt.Fprintln(os.Stderr, "\033[0;31merror:\033[0m Failed to start devcontainer")
				os.Exit(1)
			}

			time.Sleep(2 * time.Second)

			containerID, err := b.ContainerID()
			if err != nil || containerID == "" {
				fmt.Fprintln(os.Stderr, "\033[0;31merror:\033[0m Container started but couldn't find it")
				os.Exit(1)
			}

			// Create tmux window
			fmt.Println("\033[0;34m>\033[0m Setting up tmux window...")
			tmux.CreateWindow(name, containerID, b.Path)

			// Open VS Code
			if !noCode {
				fmt.Println("\033[0;34m>\033[0m Opening VS Code...")
				openVSCode(b)
			}

			// Ensure proxy
			proxy.EnsureRunning()

			fmt.Printf("\033[0;32m✓\033[0m Branch '%s' ready!\n", name)
			fmt.Println("\nAttach tmux: multi")
			fmt.Println("URLs: multi urls")
		},
	}

	cmd.Flags().StringVar(&base, "from", "main", "Base branch")
	cmd.Flags().BoolVar(&noCode, "no-code", false, "Don't open VS Code")

	return cmd
}

func startCmd() *cobra.Command {
	var noCode bool

	cmd := &cobra.Command{
		Use:   "start <branch>",
		Short: "Start stopped branch",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			name := args[0]
			b := branch.New(name)

			if !b.Exists() {
				fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m Branch '%s' not found. Create it: multi new %s\n", name, name)
				os.Exit(1)
			}

			if !b.IsManaged() {
				fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m '%s' is not a managed branch\n", name)
				os.Exit(1)
			}

			if b.IsRunning() {
				fmt.Printf("\033[1;33m!\033[0m Branch '%s' already running\n", name)
				if !tmux.WindowExists(name) {
					fmt.Println("\033[0;34m>\033[0m Adding tmux window...")
					containerID, _ := b.ContainerID()
					tmux.CreateWindow(name, containerID, b.Path)
				}
				return
			}

			fmt.Printf("\033[0;34m>\033[0m Starting branch '%s'...\n", name)

			// Ensure override config exists
			overridePath := container.GetOverrideConfigPath(b)
			if _, err := os.Stat(overridePath); os.IsNotExist(err) {
				var err error
				overridePath, err = container.GenerateOverrideConfig(b)
				if err != nil {
					fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m %v\n", err)
					os.Exit(1)
				}
				fmt.Printf("\033[0;34m>\033[0m Generated override config at %s\n", overridePath)
			}

			devCmd := exec.Command("devcontainer", "up", "--workspace-folder", b.Path, "--override-config", overridePath)
			devCmd.Stdout = os.Stdout
			devCmd.Stderr = os.Stderr
			if err := devCmd.Run(); err != nil {
				fmt.Fprintln(os.Stderr, "\033[0;31merror:\033[0m Failed to start devcontainer")
				os.Exit(1)
			}

			time.Sleep(2 * time.Second)

			containerID, err := b.ContainerID()
			if err != nil || containerID == "" {
				fmt.Fprintln(os.Stderr, "\033[0;31merror:\033[0m Container started but couldn't find it")
				os.Exit(1)
			}

			fmt.Println("\033[0;34m>\033[0m Setting up tmux window...")
			tmux.CreateWindow(name, containerID, b.Path)

			if !noCode {
				fmt.Println("\033[0;34m>\033[0m Opening VS Code...")
				openVSCode(b)
			}

			proxy.EnsureRunning()

			fmt.Printf("\033[0;32m✓\033[0m Branch '%s' running\n", name)
		},
	}

	cmd.Flags().BoolVar(&noCode, "no-code", false, "Don't open VS Code")

	return cmd
}

func stopCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "stop <branch>",
		Short: "Stop branch (keeps files)",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			name := args[0]
			b := branch.New(name)

			if !b.Exists() {
				fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m Branch '%s' not found\n", name)
				os.Exit(1)
			}

			fmt.Printf("\033[0;34m>\033[0m Stopping branch '%s'...\n", name)

			tmux.KillWindow(name)

			if containerID, err := b.ContainerID(); err == nil && containerID != "" {
				fmt.Println("\033[0;34m>\033[0m Stopping container...")
				container.StopContainer(containerID)
				container.RemoveContainer(containerID)
			}

			fmt.Printf("\033[0;32m✓\033[0m Branch '%s' stopped. Files at %s\n", name, b.Path)
		},
	}
}

func rmCmd() *cobra.Command {
	var force bool

	cmd := &cobra.Command{
		Use:   "rm <branch>",
		Short: "Remove branch entirely",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			name := args[0]
			b := branch.New(name)

			if !b.Exists() {
				fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m Branch '%s' not found\n", name)
				os.Exit(1)
			}

			if !b.IsManaged() {
				fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m '%s' is not a managed branch\n", name)
				os.Exit(1)
			}

			if !force {
				if b.HasChanges() {
					fmt.Printf("\033[1;33m!\033[0m Branch '%s' has uncommitted changes!\n", name)
				}
				fmt.Printf("This will remove branch '%s':\n", name)
				fmt.Println("  - Stop and remove container")
				fmt.Println("  - Remove tmux window")
				fmt.Printf("  - Delete %s\n", b.Path)
				fmt.Print("Proceed? [y/N] ")

				var response string
				fmt.Scanln(&response)
				if response != "y" && response != "Y" {
					fmt.Println("Aborted")
					return
				}
			}

			fmt.Printf("\033[0;34m>\033[0m Removing branch '%s'...\n", name)

			tmux.KillWindow(name)

			if containerID, err := b.ContainerID(); err == nil && containerID != "" {
				fmt.Println("\033[0;34m>\033[0m Stopping container...")
				container.StopContainer(containerID)
				container.RemoveContainer(containerID)
			}

			container.RemoveContainersByLabel(fmt.Sprintf("dark-dev-container=%s", name))

			// Remove override config
			overrideDir := filepath.Join(config.ConfigDir, "overrides", name)
			os.RemoveAll(overrideDir)

			// Remove directory
			fmt.Println("\033[0;34m>\033[0m Removing files...")
			os.RemoveAll(b.Path)

			fmt.Printf("\033[0;32m✓\033[0m Branch '%s' removed\n", name)
		},
	}

	cmd.Flags().BoolVarP(&force, "force", "f", false, "Skip confirmation")

	return cmd
}

func codeCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "code <branch>",
		Short: "Open VS Code for branch",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			name := args[0]
			b := branch.New(name)

			if !b.Exists() {
				fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m Branch '%s' not found\n", name)
				os.Exit(1)
			}

			if !b.IsRunning() {
				fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m Branch '%s' not running. Start it first: multi start %s\n", name, name)
				os.Exit(1)
			}

			fmt.Printf("\033[0;34m>\033[0m Opening VS Code for '%s'...\n", name)
			if !openVSCode(b) {
				fmt.Fprintln(os.Stderr, "\033[0;31merror:\033[0m Failed to open VS Code")
				os.Exit(1)
			}
		},
	}
}

func urlsCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "urls",
		Short: "List available URLs for branches",
		Run: func(cmd *cobra.Command, args []string) {
			branches := branch.GetManagedBranches()
			var running []*branch.Branch
			for _, b := range branches {
				if b.IsRunning() {
					running = append(running, b)
				}
			}

			if len(running) == 0 {
				fmt.Println("No running branches.")
				return
			}

			pid, isRunning := proxy.IsRunning()
			status := "not running"
			if isRunning {
				status = fmt.Sprintf("running (PID %d)", pid)
			}

			fmt.Printf("Proxy: %s\n", status)
			fmt.Printf("Port: %d\n\n", config.ProxyPort)

			fmt.Println("Running branches:")
			for _, b := range running {
				fmt.Println()
				fmt.Printf("  \033[1m%s\033[0m (ID=%d)\n", b.Name, b.InstanceID())
				fmt.Printf("    Direct:  curl -H 'Host: dark-packages.dlio.localhost' http://localhost:%d/ping\n", b.BwdPortBase())
				if isRunning {
					fmt.Printf("    Proxy:   http://dark-packages.%s.dlio.localhost:%d/ping\n", b.Name, config.ProxyPort)
				}
				fmt.Printf("    BwdServer ports: %d-%d\n", b.BwdPortBase(), b.BwdPortBase()+1)
				fmt.Printf("    Test ports: %d-%d\n", b.PortBase(), b.PortBase()+19)
			}

			if !isRunning {
				fmt.Println()
				fmt.Println("Start proxy for nice URLs: multi proxy start")
			}

			fmt.Println()
			fmt.Println("For browser access, run: multi setup-dns (one-time)")
		},
	}
}

func proxyCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "proxy <action>",
		Short: "Manage URL proxy server",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			action := args[0]

			switch action {
			case "start":
				if pid, running := proxy.IsRunning(); running {
					fmt.Printf("\033[1;33m!\033[0m Proxy already running (PID %d)\n", pid)
					return
				}

				fmt.Printf("\033[0;34m>\033[0m Starting proxy on port %d...\n", config.ProxyPort)
				pid, err := proxy.Start(config.ProxyPort, true)
				if err != nil {
					fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m %v\n", err)
					os.Exit(1)
				}
				fmt.Printf("\033[0;32m✓\033[0m Proxy started (PID %d)\n", pid)
				fmt.Println()
				fmt.Println("For browser access, run: multi setup-dns (one-time)")
				fmt.Printf("Then visit: http://dark-packages.<branch>.dlio.localhost:%d/ping\n", config.ProxyPort)

			case "stop":
				if proxy.Stop() {
					fmt.Println("\033[0;32m✓\033[0m Proxy stopped")
				} else {
					fmt.Println("\033[1;33m!\033[0m Proxy not running")
				}

			case "status":
				if pid, running := proxy.IsRunning(); running {
					fmt.Printf("Proxy running (PID %d) on port %d\n", pid, config.ProxyPort)
				} else {
					fmt.Println("Proxy not running")
				}

			case "fg":
				// Foreground mode - used by backgrounded process
				// Don't check IsRunning() here because the parent already wrote our PID to the file
				fmt.Printf("\033[0;34m>\033[0m Starting proxy on port %d (foreground)...\n", config.ProxyPort)
				proxy.Start(config.ProxyPort, false)

			default:
				fmt.Fprintf(os.Stderr, "Unknown action: %s\nUse: start, stop, status, fg\n", action)
				os.Exit(1)
			}
		},
	}

	return cmd
}

func setupDNSCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "setup-dns",
		Short: "Set up wildcard DNS for *.dlio.localhost",
		Run: func(cmd *cobra.Command, args []string) {
			if err := dns.Setup(); err != nil {
				fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m %v\n", err)
				os.Exit(1)
			}
		},
	}
}

func cmdAttach(cmd *cobra.Command, args []string) {
	// Launch TUI
	if err := tui.Run(); err != nil {
		fmt.Fprintf(os.Stderr, "\033[0;31merror:\033[0m %v\n", err)
		os.Exit(1)
	}
}

func openVSCode(b *branch.Branch) bool {
	if !b.IsRunning() {
		return false
	}

	// Use devcontainer CLI
	if _, err := exec.LookPath("devcontainer"); err == nil {
		cmd := exec.Command("devcontainer", "open", b.Path)
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		if err := cmd.Run(); err == nil {
			return true
		}
	}

	// Fallback: use code --remote
	if _, err := exec.LookPath("code"); err == nil {
		containerID, _ := b.ContainerID()
		hexID := fmt.Sprintf("%x", containerID)
		cmd := exec.Command("code", "--remote", fmt.Sprintf("attached-container+%s", hexID), "/home/dark/app")
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		return cmd.Run() == nil
	}

	fmt.Println("\033[1;33m!\033[0m Neither devcontainer CLI nor VS Code found")
	return false
}
